rmp: Genetic algorithm for slack tuning

b-chmiel opened the pull request:
New resynthesis strategy for `rmp`.
Alternative approach to https://github.com/The-OpenROAD-Project/OpenROAD/pull/8437 for optimizing ABC timing.
Utilizes a genetic algorithm with the following tunable parameters:
- population size
- mutation probability
- cross probability
- tournament size
- tournament probability

Results from regression tests:

|test|slack|execution time [s]|
|---------|------|--------------|
|[asap7 aes_genetic](https://github.com/The-OpenROAD-Project/OpenROAD/pull/8908/files#diff-a71881a40b67295fcbb984645dd09f94848fc1558a3e91c0e4401428a994009a)|+26.52|164.3|
|[asap7 aes_annealing](https://github.com/The-OpenROAD-Project/OpenROAD/blob/089d41d806190e95da38bac29965b37bcd95fd4c/src/rmp/test/aes_annealing.tcl#L1)|+14.28|52.7|

Marking as draft for a discussion, currently working on some minor implementation restructuring and algorithm parameter tuning.

Depends on https://github.com/The-OpenROAD-Project/OpenROAD/pull/8889.

calewis commented on src/rmp/README.md:
```diff
@@ -96,6 +96,11 @@ resynth_annealing
 
+### Resynth with genetic slack tuning
+
+TODO
```
Need to finish this TODO

calewis commented on src/rmp/src/slack_tuning_strategy.cpp:
```diff
@@ -0,0 +1,198 @@
+      std::swap(sol[i], sol[i + 1]);
+    } break;
+    case COUNT:
+      // unreachable
```
I would add a `__builtin_unreachable` here so you can optimize out this case.   You get slightly nicer asm https://godbolt.org/z/1ehnhqrcx, given that it actually is unreachable.   Clang and gcc both have __builtin_unreachable so I think it's very unlikely to encounter a situation where it doesn't work.

b-chmiel replied:
I used `ABSL_UNREACHABLE` from abseil. The preferable approach, `std::unreachable`, will be available since c++23.

Macro is defined here: https://github.com/abseil/abseil-cpp/blob/88c48235c8add3442800a19d5a7ee639dc242afa/absl/base/optimization.h#L207

calewis commented on src/rmp/src/genetic_strategy.cpp:
```diff
@@ -0,0 +1,171 @@
+
+  std::unordered_set<SolutionSlack::Type, HashVector> taken;
+  population.erase(
+      std::ranges::begin(std::ranges::remove_if(
```
We should be C++20 soon, but we can't use ranges till we move from 17.

calewis commented on src/rmp/src/genetic_strategy.cpp:
```diff
@@ -0,0 +1,171 @@
+    std::vector<SolutionSlack> newPopulation;
+    newPopulation.reserve(pop_size_);
+    for (int j = 0; j < pop_size_; j++) {
+      std::vector<size_t> tournament(tourn_size_);
```
If I misunderstood above then we can at least do something like this. Extract the tournament vector from the `j` loop.   Make another vector called indices that has indices 0..pop_size_-1.   Then you can do:
```
tournament_candidates.clear();
std::sample(indices.begin(), indices.end(),
                    std::back_inserter(tournament_candidates),
                    tourn_size_,
                    random_);
```
The indices will come out already sorted so you won't need to sort and erase.   The only difference is that tournament will always have size tourn_size_ since std::sample doesn't produce duplicates.

calewis commented on src/rmp/src/genetic_strategy.cpp:
```diff
@@ -0,0 +1,171 @@
+      }
+    }
+    removeDuplicates(newPopulation, logger);
+    population = newPopulation;
```
Can move here to avoid a third allocation.

calewis commented on src/rmp/src/genetic_strategy.cpp:
```diff
@@ -0,0 +1,171 @@
+      std::ranges::sort(tournament);
+      tournament.erase(std::ranges::begin(std::ranges::unique(tournament)),
+                       tournament.end());
+      std::bernoulli_distribution bern_dist{tourn_prob_};
```
We may need to switch to absl's bernoulli_distribution since the guts of std::bernoulli_distribution are implementation defined.

calewis commented on src/rmp/src/gia.cpp:
```diff
@@ -0,0 +1,426 @@
+#include "utl/unique_name.h"
+
+namespace abc {
+extern Abc_Ntk_t* Abc_NtkFromAigPhase(Aig_Man_t* pMan);
```
There isn't a header for these?

b-chmiel commented on src/rmp/src/genetic_strategy.cpp:
```diff
@@ -0,0 +1,171 @@
+      size_t res = 0;
+      for (const auto& item : sol) {
+        size_t hash = boost::hash<size_t>()(item.id);
+        // TODO set seed
```
I decided to go with absl::Hash with unordered_set approach to keep order stable and leave debugging output for each step.

b-chmiel commented on src/rmp/src/slack_tuning_strategy.cpp:
```diff
@@ -0,0 +1,198 @@
+  };
+  Move move = ADD;
+  if (sol.size() > 1) {
+    move = Move(random() % (COUNT));
```
I decided to go with absl::Uniform approach here (and in every other place where this patter occurs) to make it deterministic.

b-chmiel commented:
The only errors I'm getting right now are related to ABC imports. For instance, importing header such as `map/if/if.h` causes include cycle warning (which is there as it includes `map/if/acd/ac_wrapper.h` which includes `if.h`) and redeclaration errors in cmake build.
Not including this header causes [misc-include-cleaner](https://clang.llvm.org/extra/clang-tidy/checks/misc/include-cleaner.html) warning as those functions are not directly included in `ac_wrapper.h`. I'm also getting warnings from the inside of ABC library too.

@maliberty Should I avoid including this header altogether or ignore all problematical warnings with lint comments and compiler pragmas?

I'm also running a benchmark for https://github.com/The-OpenROAD-Project/OpenROAD/pull/8908#discussion_r2579031594 to decide which mutation approach will be best.

Besides that, I believe this PR is ready for a review.

maliberty commented:
If you don't include if.h in ac_wrapper.h, what are "those functions" that it complains about?

b-chmiel commented:
@maliberty Changes are ready, I rebased them and removed one GIA operation, as we observed it had no visible effect on slack: https://github.com/The-OpenROAD-Project/OpenROAD/pull/8908/commits/e5b9e6c5b6b0fd65e6ca7c9dbe7a06b662322a63


Currently CI is failing due old docker images in the registry after bumping of absl:
https://github.com/The-OpenROAD-Project/OpenROAD/commit/3d6152aeb786eeaf1edc459ac8ecdff46d7df2f5

error message (note old absl 20240722 instead of 20260107):
```
/usr/bin/ld: warning: libbz2.so.1, needed by /opt/or-tools/lib/libortools.so.9.14.6206, may conflict with libbz2.so.1.0

15:41:47  #9 348.4 /usr/bin/ld: /usr/local/lib/libabsl_hash.a(hash.cc.o): in function `absl::lts_20240722::hash_internal::MixingHashState::CombineLargeContiguousImpl32(unsigned long, unsigned char const*, unsigned long)':

15:41:47  #9 348.4 hash.cc:(.text+0x30): undefined reference to `absl::lts_20240722::hash_internal::CityHash32(char const*, unsigned long)'

15:41:47  #9 348.4 /usr/bin/ld: /usr/local/lib/libabsl_hash.a(hash.cc.o): in function `absl::lts_20240722::hash_internal::MixingHashState::LowLevelHashImpl(unsigned char const*, unsigned long)':

15:41:47  #9 348.4 hash.cc:(.text+0x226): undefined reference to `absl::lts_20240722::hash_internal::LowLevelHashLenGt16(void const*, unsigned long, unsigned long, unsigned long const*)'

15:41:47  #9 348.4 /usr/bin/ld: /usr/local/lib/libabsl_hash.a(hash.cc.o): in function `absl::lts_20240722::hash_internal::MixingHashState::CombineContiguousImpl(unsigned long, unsigned char const*, unsigned long, std::integral_constant<int, 4>)':

15:41:47  #9 348.4 hash.cc:(.text._ZN4absl12lts_2024072213hash_internal15MixingHashState21CombineContiguousImplEmPKhmSt17integral_constantIiLi4EE[_ZN4absl12lts_2024072213hash_internal15MixingHashState21CombineContiguousImplEmPKhmSt17integral_constantIiLi4EE]+0x63): undefined reference to `absl::lts_20240722::hash_internal::CityHash32(char const*, unsigned long)'
```

maliberty commented:
It is not a CI issue, you need to add
```
diff --git a/src/rmp/src/CMakeLists.txt b/src/rmp/src/CMakeLists.txt
index 89d7e0f288..d82e8eb509 100644
--- a/src/rmp/src/CMakeLists.txt
+++ b/src/rmp/src/CMakeLists.txt
@@ -65,6 +65,7 @@ target_link_libraries(rmp_lib
     cut
     ${ABC_LIBRARY}
     absl_hash
+    absl_city
 )
```

maliberty commented on src/rmp/src/slack_tuning_strategy.cpp:
```diff
@@ -0,0 +1,193 @@
+
+// The magic numbers are defaults from abc/src/base/abci/abc.c
+constexpr size_t SEARCH_RESIZE_ITERS = 100;
+constexpr size_t FINAL_RESIZE_ITERS = 1000;
```
Use https://google.github.io/styleguide/cppguide.html#Constant_Names

maliberty commented on src/rmp/src/slack_tuning_strategy.cpp:
```diff
@@ -0,0 +1,193 @@
+  worst_slack_ = worst_slack;
+
+  odb::dbDatabase::undoEco(block);
+  return worst_vertex;
```
How do you know worst_vertex still exists after undo?

maliberty commented on src/rmp/src/CMakeLists.txt:
```diff
@@ -48,6 +51,8 @@ target_include_directories(rmp
   PUBLIC
     ../include
+  PRIVATE
+    .
```
Unnecessary

b-chmiel commented on src/rmp/test/gcd_annealing1.ok:
Annealing and genetic use the same logic for getting random neighbors which now uses compiler-agnostic `absl::Uniform`. This way got deterministic results for both compilers. 

The annealing strategy results should not be impacted, I'll make random trial of tests for different seeds to confirm this.

b-chmiel commented:
@maliberty Resolved all comments. I'd also run experiments with different seeds for genetic and annealing slack tuning: https://github.com/The-OpenROAD-Project/OpenROAD/pull/8908#discussion_r2852088048

maliberty reviewed (approved):
Just needs conflicts resolved
